### PR TITLE
Allow a win32 window's restored size to be defined even when it is maximised or minimised

### LINKED ISSUE #13300: Window does not take into account the specified Height and Width if it is launched in the minimised state
## Describe the bug
If you set the `Window.WindowState` property in the constructor to `WindowState.Minimised`, then the window after unfolding (transition to `WindowState.Normal`) will not take into account the previously set (for example, through `ViewModel` or just in the constructor) `Window.Width` and `Window.Height.` 

## To Reproduce
1. Clone [this repo](https://github.com/flexxxxer/Avalonia.WindowSizeBug)
2. See [these lines](https://github.com/flexxxxer/Avalonia.WindowSizeBug/blob/master/Avalonia.WindowSizeBug/Views/MainWindow.axaml.cs#L9-L12): `MainWindow.Width`, `.Height` and `.WindowState` are set in the constructor.
3. Build and run app: window will be in minimized state.
4. After opening an application window (e.g. clicking in the taskbar), the application window will have a different size than the one specified. 

## Expected behavior
Application window size after application opening corresponds to the values set in the constructor.

## Screenshots
https://github.com/AvaloniaUI/Avalonia/assets/44833369/1ae5908c-2bbf-4858-bff2-139fb2752030

## Environment 

- OS: Windows 11
- Avalonia-Version: reproduced on 0.10.5, 0.10.10, 0.10.21, 11.0.0, 11.0.1, 11.0.2, 11.0.3, 11.0.4, 11.0.5

### PR DESCRIPTION
**Pull request by TomEdwardsEnscape**
This PR rewrites `WindowImpl.Resize` to use `SetWindowPlacement`, which allows us to set the size of the window even when it is minimised or maximised.

## Breaking changes
None

## Obsoletions / Deprecations
None

## Fixed issues
Fixes #13300

### CODE REVIEW

**Comment by hez2010** [comment]
I don't think this is the correct fix.
A minimized window is not an actual window but a placeholder without any size information on it.
If we want to change the size before the window shows up, I will rather expect to save the size information and resize it when users restore the window state to normal.

**Comment by TomEdwardsEnscape** [comment]
> A minimized window is not an actual window but a placeholder without any size information on it.

It might be a placeholder, but nevertheless the size of a minimised window is stored by Windows and is sent back to the application as a `WM_SIZE` message when that window is restored.

[This is the behaviour that Avalonia expects](https://github.com/AvaloniaUI/Avalonia/blob/0e9ce1522391ab91a2950848c74ebeb08fff4108/src/Windows/Avalonia.Win32/WindowImpl.cs#L239-L247), so I think that removing the inconsistency between minimised and non-minimised windows is a good approach.

> If we want to change the size before the window shows up, I will rather expect to save the size information and resize it when users restore the window state to normal. But I think this should be handled by the app developer, not the framework itself.

This is OK in principle, but it doesn't map well onto the Avalonia API, in which windows have public `Width` and `Height` properties that can be read and written at any time. Should we start throwing exceptions when those properties are set while the window is minimised?

**Comment by hez2010** [comment]
> This is OK in principle, but it doesn't map well onto the Avalonia API, in which windows have public `Width` and `Height` properties that can be read and written at any time. Should we start throwing exceptions when those properties are set while the window is minimised?

The behavior in other win32 applications is just to return the window size before it was last minimized, and any resize operation is ignored after the window was minimized.

**Comment by TomEdwardsEnscape** [comment]
Avalonia doesn't ignore the resize. It accepts it the new value for its own window object without transferring it to the OS, which is inconsistent. It should either apply the new value in both places, or in neither (i.e. throw an exception).

**Comment by thevortexcloud** [comment]
It should either outright reject the change or buffer it from the sound of things. Quickly making it visible then invisible feels like a hack to me. If you do go with rejecting the change, you probably also need to add a new property/method that says if the window can be resized. `CanResize` is currently settable by user code and indicates if the end user is allowed to resize the window.

**Comment by TomEdwardsEnscape** [comment]
I don't think the window would ever appear, because we are executing on the UI thread and thus blocking it. But yeah, buffering the value is the (more complicated) alternative if we don't like re-minimising.

I'm not a fan of blocking changes to the properties because these limitations may not exist on other platforms.

**Comment by StefanKoell** [comment]
Coming from WinForms, I know there's a concept called "RestoreBounds" on a form:
https://learn.microsoft.com/en-us/dotnet/api/system.windows.window.restorebounds?view=windowsdesktop-8.0

`A Rect that specifies the size and location of a window before being either minimized or maximized.`

Maybe something like this would be a good approach.

**Comment by jp2masa** [comment]
Can't this be fixed using [`SetWindowPlacement`](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setwindowplacement)?

**Comment by TomEdwardsEnscape** [comment]
> Can't this be fixed using [`SetWindowPlacement`](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setwindowplacement)?

Well well well:

> WPF_SETMINPOSITION: The coordinates of the minimized window may be specified. This flag must be specified if the coordinates are set in the ptMinPosition member.

That's our solution, and is apparently also what WPF does to handle this exact scenario.

**Comment by TomEdwardsEnscape** [comment]
`SetWindowPlacement` worked! It supports setting the window size of a minimised window without restoring it.

File: src/Windows/Avalonia.Win32/WindowImpl.cs [hunk 1]
```diff
@@ -1426,6 +1450,23 @@ private static void EnableCloseButton(IntPtr hwnd)
                            MF_BYCOMMAND | MF_ENABLED);
         }
 
+        private RECT ClientRectToWindowRect(RECT clientRect, WindowStyles? styleOverride = null, WindowStyles? extendedStyleOverride = null)
```
Comment (TomEdwardsEnscape) [inline src/Windows/Avalonia.Win32/WindowImpl.cs:1453]: There are several other uses of `AdjustWindowRectEx` which could be redirected to this method. These other locations currently don't account for per-window DPI, so using this method may fix some bugs. But I decided not to make unnecessary changes in this PR.

File: src/Windows/Avalonia.Win32/WindowImpl.cs [hunk 2]
```diff
@@ -556,29 +561,59 @@ public void SetMinMaxSize(Size minSize, Size maxSize)
...
+            {
+                // Don't update our window position if the client size is already correct. This leads to Windows updating our
+                // "normal position" (i.e. restored bounds) to match our maximised or areo snap size, which is incorrect behaviour.
+                // We only want to proceed with this method if the new size is coming from Avalonia.
+                return;
+            }
 
-            // do comparison after scaling to avoid rounding issues
-            if (requestedClientWidth != clientRect.Width || requestedClientHeight != clientRect.Height)
+            if (_lastWindowState == WindowState.FullScreen)
```
Comment (TomEdwardsEnscape) [inline src/Windows/Avalonia.Win32/WindowImpl.cs:576]: We still can't resize a fullscreen window, but I don't think this will ever be supported for the reasons outlined in the comment here. This is not a regression, it's the same behaviour as on master.

**Comment by TomEdwardsEnscape** [comment]
There are still a few integration tests failing, but I don't understand why. Appium can't find the elements it's looking for. Everything is fine when I step replicate these tests manually.

```
[xUnit.net 00:03:51.34]     Avalonia.IntegrationTests.Appium.WindowTests.WindowState(size: 400, 300, mode: NonOwned, state: FullScreen, canResize: False) [FAIL]
  Failed Avalonia.IntegrationTests.Appium.WindowTests.WindowState(size: 400, 300, mode: NonOwned, state: FullScreen, canResize: False) [4 s]
  Error Message:
   The collection was expected to contain a single element, but it was empty.
  Stack Trace:
     at Avalonia.IntegrationTests.Appium.ElementExtensions.OpenWindowWithClick(AppiumWebElement element) in D:\a\1\s\tests\Avalonia.IntegrationTests.Appium\ElementExtensions.cs:line 109
   at Avalonia.IntegrationTests.Appium.WindowTests.OpenWindow(Nullable`1 size, ShowWindowMode mode, WindowStartupLocation location, WindowState state, Boolean canResize, Boolean extendClientArea) in D:\a\1\s\tests\Avalonia.IntegrationTests.Appium\WindowTests.cs:line 413
   at Avalonia.IntegrationTests.Appium.WindowTests.WindowState(Nullable`1 size, ShowWindowMode mode, WindowState state, Boolean canResize) in D:\a\1\s\tests\Avalonia.IntegrationTests.Appium\WindowTests.cs:line 66

[xUnit.net 00:05:24.17]     Avalonia.IntegrationTests.Appium.WindowTests.Window_Has_Disabled_Maximize_Button_When_CanResize_Is_False(mode: Owned, extendClientArea: False) [FAIL]
  Failed Avalonia.IntegrationTests.Appium.WindowTests.Window_Has_Disabled_Maximize_Button_When_CanResize_Is_False(mode: Owned, extendClientArea: False) [7 s]
  Error Message:
   OpenQA.Selenium.WebDriverException : An element could not be located on the page using the given search parameters.
  Stack Trace:
     at OpenQA.Selenium.Remote.RemoteWebDriver.UnpackAndThrowOnError(Response errorResponse)
   at OpenQA.Selenium.Remote.RemoteWebDriver.Execute(String driverCommandToExecute, Dictionary`2 parameters)
   at OpenQA.Selenium.Remote.RemoteWebDriver.FindElement(String mechanism, String value)
   at OpenQA.Selenium.Remote.RemoteWebDriver.FindElementByXPath(String xpath)
   at OpenQA.Selenium.Appium.AppiumDriver`1.FindElementByXPath(String xpath)
   at Avalonia.IntegrationTests.Appium.WindowTests.GetWindow(String identifier) in D:\a\1\s\tests\Avalonia.IntegrationTests.Appium\WindowTests.cs:line 427
   at Avalonia.IntegrationTests.Appium.WindowTests.Window_Has_Disabled_Maximize_Button_When_CanResize_Is_False(ShowWindowMode mode, Boolean extendClientArea) in D:\a\1\s\tests\Avalonia.IntegrationTests.Appium\WindowTests.cs:line 271
```

**Comment by TomEdwardsEnscape** [comment]
@emmauss this is ready for review again.

**Review by maxkatz6 — APPROVED** [review]

**Comment by maxkatz6** [comment]
LGTM, just want to have @emmauss opinion as well, as it might affect XPF.

**Comment by emmauss** [comment]
Could you rebase your branch?

**Review by emmauss — APPROVED** [review]

**Comment by maxkatz6** [comment]
@TomEdwardsEnscape it looks like there are conflicts. It cannot be automatically resolved, and we can't push to your remote branch.
